Adding cone primitives.

bperseghetti opened the pull request:
# 🦟 Bug fix

## Summary
This helps add the missing cone geometry for primitive/basic parametric shapes:

![conetopple](https://github.com/gazebosim/gz-math/assets/10233412/5fd8f1a1-3a77-4e61-95d5-f053389cd908)
![cone](https://github.com/gazebosim/gz-math/assets/10233412/1c516775-7adb-4318-9c6a-0c09a746a3b0)

And is also valuable for visualizations of emitters/source that typically have conic-based spread as seen in this acoustic attack on an IMU by showing the affected area:

![drone_attack](https://github.com/gazebosim/gz-rendering/assets/10233412/7a6b0dfa-8ad6-42c1-83bc-8385ccc4c81a)

Associated PRs:
- https://github.com/gazebosim/gz-gui/pull/621
- https://github.com/gazebosim/gz-math/pull/594
- https://github.com/gazebosim/gz-msgs/pull/442
- https://github.com/gazebosim/gz-physics/pull/639
- https://github.com/gazebosim/gz-rendering/pull/1003
- https://github.com/gazebosim/gz-sim/pull/2410
- https://github.com/gazebosim/sdformat/pull/1418

## Checklist
- [x] Signed all commits for DCO
- [x] Added tests
- [x] Updated documentation (as needed)
- [x] Updated migration guide (as needed)
- [x] Consider updating Python bindings (if the library has them)
- [x] `codecheck` passed (See [contributing](https://gazebosim.org/docs/all/contributing#contributing-code))
- [x] All tests passed (See [test coverage](https://gazebosim.org/docs/all/contributing#test-coverage))
- [x] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Agazebosim+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

azeey commented on src/python_pybind11/src/Helpers.cc:
```diff
@@ -180,7 +192,28 @@ void defineMathHelpers(py::module &m)
         "Compute box volume")
    .def("gz_box_volume_v",
         &BoxVolumeV,
+<<<<<<< HEAD
```
This is causing a build failure

bperseghetti replied:
Dang it, that one must have also slipped through on those cherry-pick conflicts, thought I had got all ~15 of them... Fixed now.

scpeters commented on src/python_pybind11/test/Cone_TEST.py:
```diff
@@ -0,0 +1,126 @@
+import unittest
+
+import gz
+from gz.math7 import Coned, MassMatrix3d, Material, Quaterniond
```
```suggestion
from gz.math8 import Coned, MassMatrix3d, Material, Quaterniond
```

bperseghetti commented:
Yay, I finally figured out how to bump math7->math8 to get it past CI :man_facepalming:... :melting_face:

ahcorde commented on include/gz/math/Cone.hh:
```diff
@@ -0,0 +1,193 @@
+      /// \param[in] _radius Radius of the cone.
+      /// \param[in] _rotOffset Rotational offset of the cone.
+      public: Cone(const Precision _length, const Precision _radius,
+                  const Quaternion<Precision> &_rotOffset =
+                  Quaternion<Precision>::Identity);
```
```suggestion
                   const Quaternion<Precision> &_rotOffset =
                   Quaternion<Precision>::Identity);
```

ahcorde commented on include/gz/math/Cone.hh:
```diff
@@ -0,0 +1,193 @@
+      /// \param[in] _mat Material property for the cone.
+      /// \param[in] _rotOffset Rotational offset of the cone.
+      public: Cone(const Precision _length, const Precision _radius,
+                  const Material &_mat,
+                  const Quaternion<Precision> &_rotOffset =
+                  Quaternion<Precision>::Identity);
```
```suggestion
                   const Material &_mat,
                   const Quaternion<Precision> &_rotOffset =
                   Quaternion<Precision>::Identity);
```

ahcorde commented on include/gz/math/Cone.hh:
```diff
@@ -0,0 +1,193 @@
+      public: Quaternion<Precision> RotationalOffset() const;
+
+      /// \brief Set the rotation offset.
+      /// \param[in] _rotOffset
```
add description
```suggestion
      /// \param[in] _rotOffset
```

bperseghetti replied:
This was copied directly from all the other ones, IE this doesn't even have it:
https://github.com/gazebosim/gz-math/blob/2f1e049bc6937efab3fd7a1d136978fc3339cf30/include/gz/math/Cylinder.hh#L94-L97

But I'll add  "Rotational offset quaternion"

ahcorde commented on include/gz/math/MassMatrix3.hh:
```diff
@@ -930,6 +930,83 @@ namespace gz
+      /// \param[in] _radius Radius of cone.
+      /// \param[in] _rot Rotational offset of equivalent cone.
+      /// \return True if inertial properties were set successfully.
+      public: bool SetFromConeZ(const Material &_mat,
+                                    const T _length,
+                                    const T _radius,
+                            const Quaternion<T> &_rot = Quaternion<T>::Identity)
```
```suggestion
      public: bool SetFromConeZ(const Material &_mat,
                                const T _length,
                                const T _radius,
                                const Quaternion<T> &_rot = Quaternion<T>::Identity)
```

ahcorde commented on include/gz/math/MassMatrix3.hh:
```diff
@@ -930,6 +930,83 @@ namespace gz
+      /// \param[in] _radius Radius of cone.
+      /// \param[in] _rot Rotational offset of equivalent cone.
+      /// \return True if inertial properties were set successfully.
+      public: bool SetFromConeZ(const T _mass,
+                                    const T _length,
+                                    const T _radius,
+                            const Quaternion<T> &_rot = Quaternion<T>::Identity)
```
```suggestion
      public: bool SetFromConeZ(const T _mass,
                                const T _length,
                                const T _radius,
                                const Quaternion<T> &_rot = Quaternion<T>::Identity)
```

ahcorde commented on include/gz/math/MassMatrix3.hh:
```diff
@@ -930,6 +930,83 @@ namespace gz
+      /// \param[in] _radius Radius of cone.
+      /// \param[in] _rot Rotational offset of equivalent cone.
+      /// \return True if inertial properties were set successfully.
+      public: bool SetFromConeZ(const T _length,
+                                    const T _radius,
+                                    const Quaternion<T> &_rot)
```
```suggestion
      public: bool SetFromConeZ(const T _length,
                                const T _radius,
                                const Quaternion<T> &_rot)
```

ahcorde commented on include/gz/math/MassMatrix3.hh:
```diff
@@ -930,6 +930,83 @@ namespace gz
+        }
+
+        // Diagonal matrix L with principal moments
+        T radius2 = std::pow(_radius, 2);
```
include `<cmath>`

ahcorde commented on src/ruby/MassMatrix3.i:
```diff
@@ -99,10 +99,21 @@ namespace gz
                                     const T _radius,
                                     const Quaternion<T> &_rot);
+      public: bool SetFromConeZ(const Material &_mat,
+                                    const T _length,
+                                    const T _radius,
+                                    const Quaternion<T> &_rot = Quaternion<T>::Identity);
+      public: bool SetFromConeZ(const T _mass,
+                                    const T _length,
+                                    const T _radius,
+                                    const Quaternion<T> &_rot = Quaternion<T>::Identity);
+      public: bool SetFromConeZ(const T _length,
+                                    const T _radius,
+                                    const Quaternion<T> &_rot);
```
```suggestion
                                const T _length,
                                const T _radius,
                                const Quaternion<T> &_rot = Quaternion<T>::Identity);
      public: bool SetFromConeZ(const T _mass,
                                const T _length,
                                const T _radius,
                                const Quaternion<T> &_rot = Quaternion<T>::Identity);
      public: bool SetFromConeZ(const T _length,
                                const T _radius,
                                const Quaternion<T> &_rot);
```

ahcorde commented on include/gz/math/Cone.hh:
```diff
@@ -0,0 +1,193 @@
+      /// \param[in] _length Length of the cone.
+      /// \param[in] _radius Radius of the cone.
+      /// \param[in] _rotOffset Rotational offset of the cone.
+      public: Cone(const Precision _length, const Precision _radius,
```
I think we should remove all `const` from `Precision`. This is a primitive, and we don't use `const` with basic primitives. @scpeters or @azeey, thoughts ?

bperseghetti replied:
@ahcorde I think that is just a copy/paste from Cylinder with minor changes FYI:
https://github.com/gazebosim/gz-math/blob/2f1e049bc6937efab3fd7a1d136978fc3339cf30/include/gz/math/Cylinder.hh#L51-L157

bperseghetti replied:
And if we change it in Cone we would probably need to change it in a lot of other places?

bperseghetti replied:
I also just pushed a PR to fix the linting issues I think popped up from the last "batch suggestion apply". Should all be green after that.

scpeters replied:
address the `const` issues in a separate PR where we can make multiple changes in a consistent way

bperseghetti replied:
@scpeters any other changes desired/needed on this particular PR?

scpeters commented on src/Cone_TEST.cc:
```diff
@@ -0,0 +1,147 @@
+{
+  double mass = 1.0;
+  math::Coned cone(1.0, 0.001);
+  double expectedVolume = (GZ_PI * std::pow(0.001, 2) * 1.0);
```
```suggestion
  double expectedVolume = (GZ_PI * std::pow(0.001, 2) * 1.0 / 3.0);
```

scpeters commented on include/gz/math/MassMatrix3.hh:
```diff
@@ -930,6 +931,85 @@ namespace gz
+        {
+          return false;
+        }
+        T volume = GZ_PI * _radius * _radius * _length;
```
```suggestion
        T volume = GZ_PI * _radius * _radius * _length / 3.0;
```

scpeters commented on include/gz/math/detail/Cone.hh:
```diff
@@ -0,0 +1,172 @@
+T Cone<T>::Volume() const
+{
+  return GZ_PI * std::pow(this->radius, 2) *
+         this->length;
```
```suggestion
         this->length / 3.0;
```

scpeters commented on src/python_pybind11/test/Cone_TEST.py:
```diff
@@ -0,0 +1,126 @@
+    def test_volume_and_density(self):
+        mass = 1.0
+        cone = Coned(1.0, 0.001)
+        expectedVolume = (math.pi * math.pow(0.001, 2) * 1.0)
```
```suggestion
        expectedVolume = (math.pi * math.pow(0.001, 2) * 1.0 / 3.0)
```

scpeters reviewed (commented):
I checked out your code and reviewed a bit of it on a flight yesterday but didn't have a chance to post my comments until now. It looks like you've mainly copied the Cylinder class, which makes sense, but the Volume computation needs to be updated in a few places.

I haven't yet reviewed the moment of inertia calculations.

bperseghetti commented:
> I checked out your code and reviewed a bit of it on a flight yesterday but didn't have a chance to post my comments until now. It looks like you've mainly copied the Cylinder class, which makes sense, but the Volume computation needs to be updated in a few places.
> 
> I haven't yet reviewed the moment of inertia calculations.

@scpeters yeah, I think I had did the volume properly everywhere except for some of the tests I had added around 4AM this morning while on "awake kiddo" duty. 
According to wolfram here is the inertia we use for cylinder at center of mass: https://resources.wolframcloud.com/FormulaRepository/resources/Moment-of-Inertia-of-a-Cylinder
![image](https://github.com/gazebosim/gz-math/assets/10233412/f2910462-3e52-407d-b5e9-74db41428a5d)

And according to wolfram here it is for the cone at center of mass that I added to the code:
https://resources.wolframcloud.com/FormulaRepository/resources/Moment-of-Inertia-of-a-Cone
![image](https://github.com/gazebosim/gz-math/assets/10233412/5594f767-9166-40fc-9c8d-8f3f6cccabeb)

bperseghetti commented:
> > > I checked out your code and reviewed a bit of it on a flight yesterday but didn't have a chance to post my comments until now. It looks like you've mainly copied the Cylinder class, which makes sense, but the Volume computation needs to be updated in a few places.
> > > I haven't yet reviewed the moment of inertia calculations.
> > 
> > 
> > @scpeters yeah, I think I had did the volume properly everywhere except for some of the tests I had added around 4AM this morning while on "awake kiddo" duty. According to wolfram here is the inertia we use for cylinder at center of mass: https://resources.wolframcloud.com/FormulaRepository/resources/Moment-of-Inertia-of-a-Cylinder ![image](https://private-user-images.githubusercontent.com/10233412/331673134-f2910462-3e52-407d-b5e9-74db41428a5d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTU5NjU0NDIsIm5iZiI6MTcxNTk2NTE0MiwicGF0aCI6Ii8xMDIzMzQxMi8zMzE2NzMxMzQtZjI5MTA0NjItM2U1Mi00MDdkLWI1ZTktNzRkYjQxNDI4YTVkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTE3VDE2NTkwMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPThmODY2NmI5ZTY4NWNjNGUzNWQzMGI1ZDA1Mjc0NjA5ZWMzMGFmNzA5ZWJiY2E3NTRiNmJjZWZiNDUwZTY1ZWUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.XToCWo1IAhoGVVblRo2OHhsxvL1oqTMdh52pe8aSEBk)
> > And according to wolfram here it is for the cone at center of mass that I added to the code: https://resources.wolframcloud.com/FormulaRepository/resources/Moment-of-Inertia-of-a-Cone ![image](https://private-user-images.githubusercontent.com/10233412/331673060-5594f767-9166-40fc-9c8d-8f3f6cccabeb.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTU5NjU0NDIsIm5iZiI6MTcxNTk2NTE0MiwicGF0aCI6Ii8xMDIzMzQxMi8zMzE2NzMwNjAtNTU5NGY3NjctOTE2Ni00MGZjLTljOGQtOGYzZjZjY2NhYmViLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTE3VDE2NTkwMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWVjMWRhMzQzZjE4MDZiNDdjMGY2NjdhZGVlZGNhZmQzMTRjYTFkYjI1MTljYTliOWY3NGJjNTk4MDM1MzYwYzkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.WzENqz8LVif2x2n0S7B49001U7xHQMpL3IVHL-CVQT4)
> 
> Great those inertia values relative to the center of mass look good to me. The tricky thing about cone inertial properties compared to the other simple shapes that we support is that a cone's center of mass is not at the geometric center of the shape. So when we specify model properties and geometry in SDFormat, if a link has a single shape with uniform density, we can set the link pose and then can use the default identity poses for the collision, visual, and inertial tags. With a cone, we need to be explicit about where the shape origin is defined. If we use the geometric center, then a link with uniform density cone could still use the identity pose for collision and visual but would need the inertial pose to be set with the center of mass location.
> 
> Most of that is a concern for sdformat, which I will address in review of [gazebosim/sdformat#1418](https://github.com/gazebosim/sdformat/pull/1418), but I think the `Cone` class could use an additional API to help with this. All the other shapes have a `MassMatrix()` method to get inertial properties at the center of mass, but I think the `Cone` could additionally use an API that returns an `Inertial` object with the center of mass location encoded. I'm not going to block this PR on adding that API, but I think I will ask for it to get added to `gz::math::Cone` in the course of reviewing [gazebosim/sdformat#1418](https://github.com/gazebosim/sdformat/pull/1418).

I believe that for how cone is defined the center of mass is `length / 4.0`

bperseghetti commented:
@scpeters @azeey @ahcorde  Awesome! So I think the next one to go is gz-msgs? https://github.com/gazebosim/gz-msgs/pull/442

And once that is in the nightlies should update for the others to start working in CI?
